sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriterV2.scala

```diff
@@ -167,6 +173,59 @@ final class DataFrameWriterV2[T] private[sql](table: String, ds: Dataset[T])
     runCommand(overwrite)
   }
 
+  def on(condition: Column): MergeIntoWriter[T] = {
+    this.on = Some(condition)
+    this
+  }
+
+  def whenMatched(): WhenMatched[T] = {
```

Comment:
in SQL, users can't put `WHEN MATCHED` tiwce. Can we enforce it in the scala API as well? e.g.
```
df.mergeInto...
.whenMatched...
.whenMatched... // runtime error
```